observer/observer.go

```diff
@@ -152,14 +158,57 @@ func (o *Observer) Watch(files []string) error {
 	return nil
 }
 
-// handleEvent handle an event.
-func (o *Observer) handleEvent(event interface{}) {
-	// Run all listeners for this event.
+// SetBufferDuration set the buffer damping duration
+func (o *Observer) SetBufferDuration(d time.Duration) {
+	// Create the buffer mutex, if missing.
+	if o.bufferMux == nil {
+		o.bufferMux = &sync.Mutex{}
+	}
+
+	// Set the buffer duration.
+	o.bufferDuration = d
+}
+
+// sendEvent send one or more events to the observer listeners.
+func (o *Observer) sendEvent(event interface{}) {
 	for _, listener := range o.listeners {
 		go listener(event)
 	}
 }
 
+// handleEvent handle an event.
+func (o *Observer) handleEvent(event interface{}) {
+	// If we do not buffer events, just send this event now.
+	if o.bufferDuration == 0 {
+		o.sendEvent(event)
+		return
+	}
+
+	// Add new event to the event buffer
+	o.bufferMux.Lock()
+	o.bufferEvents = append(o.bufferEvents, event)
+	o.bufferMux.Unlock()
+
+	// If this is the first event, set a timeout function.
+	if o.bufferTimer == nil {
```

Comment:
I think there might be a race condition here.
Lets say you have two go routines that are executing `handleEvent` both are about to execute this line.
go-routine #1 will see that buffer is not nil and jump to the block but before executing the first line in the block the CPU stops it and runs go-routine #2 which reads this and jumps to this block because `o.bufferTimer` hasn't been set yet.

Perhaps adding:
```
o.bufferTimerMutex.Lock()
defer o.bufferTimerMutex.Unlock()
```
solves this problem.

WDYT?

Reply:
Thanks !